`mdbook` fails when used with external repository

Hello! I have been trying to use `mdbook` with sources coming from an external repository and I'm find that I get

> [ERROR] (`mdbook::utils`): Error: Couldn't open `SUMMARY.md` in `".../repro-external-sources_/external/content+/src"`

errors back. I had Gemini reproduce the problem in https://github.com/mgeisler/rules-rust-mdbook-external-sources and I hope this is useful for you.

### Description

When any part of an `mdbook` target (either the `book.toml` or the source files) comes from an _external repository_, the build fails. This is because the rule's staging logic and its invocation logic handle repository boundaries inconsistently.

### Reproduction steps

Please see https://github.com/mgeisler/rules-rust-mdbook-external-sources

### Additional context

From Gemini: The failure stems from a mismatch between the Starlark rule definition and the Rust `process_wrapper` regarding how external repository paths are resolved.

1. **Staging (`process_wrapper.rs`):** The `MdBookBuild` action uses a Rust wrapper to stage files into a shadow work directory (`<target>_`). This wrapper receives an input manifest where destinations are derived from `file.short_path`. In Bazel, the `short_path` for files in an external repository starts with `../` (e.g., `../content+/src/SUMMARY.md`). This causes the files to be staged **outside** the shadow root (as siblings to it).

2. **Invocation (`mdbook.bzl`):** The `_mdbook_impl` function in `mdbook.bzl` tells `mdbook` where the book's root is using `${pwd}/{}` formatted with `ctx.file.book.dirname`. For external repositories, `dirname` (based on `file.path`) starts with `external/` (e.g., `external/content+/`).

3. **The Mismatch:** `mdbook` is invoked with a path like `.../repro-external-sources_/external/content+/`, but the files were actually staged by the wrapper at `.../repro-external-sources_/../content+/`.

### Impact

I'm likely going to drive `mdbook` via a simple `genrule` instead.

### Bazel and rules_rust version

- Bazel 9.1.0
- `rules_rust` 0.70.0

--- a/extensions/mdbook/private/mdbook.bzl
+++ b/extensions/mdbook/private/mdbook.bzl
@@ -10,7 +10,15 @@ MdBookInfo = provider(
 )
 
 def _map_inputs(file):
-    return "{}={}".format(file.path, file.short_path)
+    dest = file.short_path
+    if dest.startswith("../"):
+        # External repositories have short_paths starting with '../'.
+        # We need them to be staged at 'external/' within our shadow
+        # directory to match how 'file.path' (and thus 'file.dirname')
+        # refers to them.
+        dest = "external/" + dest.removeprefix("../")
+
+    return "{}={}".format(file.path, dest)
 
 def _mdbook_impl(ctx):
     output = ctx.actions.declare_directory(ctx.label.name)

--- a/extensions/mdbook/private/process_wrapper.rs
+++ b/extensions/mdbook/private/process_wrapper.rs
@@ -73,6 +73,17 @@ fn generate_work_dir(output_dir: &Path, inputs_manifest: &BTreeMap<PathBuf, Path
 
     for (src, dest) in inputs_manifest.iter() {
         let abs_dest = workdir.join(dest);
+
+        // Ensure the destination is within the workdir
+        if !abs_dest.starts_with(&workdir) {
+            panic!(
+                "Attempted to stage file outside of workdir: {} -> {} (dest: {})",
+                src.display(),
+                abs_dest.display(),
+                dest.display()
+            );
+        }
+
         fs::create_dir_all(abs_dest.parent().unwrap()).unwrap();
         fs::copy(src, &abs_dest).unwrap_or_else(|e| {
             panic!(

--- a/extensions/mdbook/test/external_srcs/BUILD.bazel
+++ b/extensions/mdbook/test/external_srcs/BUILD.bazel
@@ -1,0 +1,35 @@
+load("@bazel_skylib//rules:build_test.bzl", "build_test")
+load("@rules_rust//rust:defs.bzl", "rust_test")
+load("//:defs.bzl", "mdbook")
+
+mdbook(
+    name = "external_srcs",
+    srcs = ["//test/external_srcs/content:srcs"],
+    book = "//test/external_srcs/content:book.toml",
+)
+
+mdbook(
+    name = "local_book_mixed",
+    srcs = ["//test/external_srcs/content:srcs"],
+    book = "//test/external_srcs/local_book_mixed:book.toml",
+)
+
+rust_test(
+    name = "external_srcs_test",
+    srcs = ["external_srcs_test.rs"],
+    data = [
+        ":external_srcs",
+        ":local_book_mixed",
+    ],
+    edition = "2021",
+    rustc_env = {
+        "MDBOOK_EXTERNAL_SRCS_OUTPUT": "$(rlocationpath :external_srcs)",
+        "MDBOOK_LOCAL_BOOK_MIXED_OUTPUT": "$(rlocationpath :local_book_mixed)",
+    },
+    deps = ["@rules_rust//rust/runfiles"],
+)
+
+build_test(
+    name = "external_srcs_build_test",
+    targets = [":external_srcs"],
+)

--- a/extensions/mdbook/test/external_srcs/content/BUILD.bazel
+++ b/extensions/mdbook/test/external_srcs/content/BUILD.bazel
@@ -1,0 +1,7 @@
+exports_files(["book.toml"])
+
+filegroup(
+    name = "srcs",
+    srcs = glob(["src/**/*.md"]),
+    visibility = ["//visibility:public"],
+)

--- a/extensions/mdbook/test/external_srcs/content/book.toml
+++ b/extensions/mdbook/test/external_srcs/content/book.toml
@@ -1,0 +1,5 @@
+[book]
+authors = ["Test Author"]
+language = "en"
+src = "src"
+title = "External Book"

--- a/extensions/mdbook/test/external_srcs/content/src/SUMMARY.md
+++ b/extensions/mdbook/test/external_srcs/content/src/SUMMARY.md
@@ -1,0 +1,3 @@
+# Summary
+
+[Test Chapter](test.md)

--- a/extensions/mdbook/test/external_srcs/content/src/test.md
+++ b/extensions/mdbook/test/external_srcs/content/src/test.md
@@ -1,0 +1,3 @@
+# Test Chapter
+
+This is a test.

--- a/extensions/mdbook/test/external_srcs/external_srcs_test.rs
+++ b/extensions/mdbook/test/external_srcs/external_srcs_test.rs
@@ -1,0 +1,24 @@
+use runfiles::{rlocation, Runfiles};
+use std::fs;
+
+#[test]
+fn test_external_srcs_output() {
+    let r = Runfiles::create().unwrap();
+
+    let dir = rlocation!(r, env!("MDBOOK_EXTERNAL_SRCS_OUTPUT")).unwrap();
+
+    let test_chapter = dir.join("test.html");
+    let content = fs::read_to_string(test_chapter).unwrap();
+    assert!(content.contains("This is a test."));
+}
+
+#[test]
+fn test_local_book_mixed_output() {
+    let r = Runfiles::create().unwrap();
+
+    let dir = rlocation!(r, env!("MDBOOK_LOCAL_BOOK_MIXED_OUTPUT")).unwrap();
+
+    let test_chapter = dir.join("test.html");
+    let content = fs::read_to_string(test_chapter).unwrap();
+    assert!(content.contains("This is a test."));
+}

--- a/extensions/mdbook/test/external_srcs/local_book_mixed/BUILD.bazel
+++ b/extensions/mdbook/test/external_srcs/local_book_mixed/BUILD.bazel
@@ -1,0 +1,1 @@
+exports_files(["book.toml"])

--- a/extensions/mdbook/test/external_srcs/local_book_mixed/book.toml
+++ b/extensions/mdbook/test/external_srcs/local_book_mixed/book.toml
@@ -1,0 +1,5 @@
+[book]
+authors = ["Test Author"]
+language = "en"
+src = "../content/src"
+title = "Local Book Mixed"

